src/Features/Core/Portable/AddDebuggerDisplay/AbstractAddDebuggerDisplayCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,181 @@
+            }
+        }
+
+        private static async Task<TTypeDeclarationSyntax> GetRelevantTypeFromMethodAsync(CodeRefactoringContext context)
+        {
+            var method = await context.TryGetRelevantNodeAsync<TMethodDeclarationSyntax>().ConfigureAwait(false);
+            if (method != null)
+            {
+                var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+                var methodSymbol = (IMethodSymbol)semanticModel.GetDeclaredSymbol(method);
+
+                if (IsToStringOverride(methodSymbol) || IsDebuggerDisplayMethod(methodSymbol))
+                {
+                    return method.FirstAncestorOrSelf<TTypeDeclarationSyntax>();
+                }
+            }
+
+            return null;
+        }
+
+        private static bool IsToStringOverride(IMethodSymbol methodSymbol)
+        {
+            return methodSymbol is
+            {
+                Arity: 0,
+                Parameters: { IsEmpty: true },
+                Name: nameof(ToString),
+                IsOverride: true
+            };
+        }
+
+        private static bool IsDebuggerDisplayMethod(IMethodSymbol methodSymbol)
+        {
+            return methodSymbol is
+            {
+                Arity: 0,
+                Parameters: { IsEmpty: true },
+                Name: DebuggerDisplayMethodName
+            };
+        }
+
+        private static bool IsClassOrStruct(ITypeSymbol typeSymbol)
+        {
+            switch (typeSymbol.TypeKind)
+            {
+                case TypeKind.Class:
+                case TypeKind.Struct:
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        private static bool HasDebuggerDisplayAttribute(ITypeSymbol typeSymbol, Compilation compilation)
+        {
+            return typeSymbol.GetAttributes()
+                .Select(data => data.AttributeClass)
+                .Contains(compilation.GetTypeByMetadataName("System.Diagnostics.DebuggerDisplayAttribute"));
+        }
+
+        private async Task<Document> ApplyAsync(Document document, TTypeDeclarationSyntax type, CancellationToken cancellationToken)
+        {
+            var syntaxRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var symbolicModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            var newAttribute = generator
+                .Attribute("System.Diagnostics.DebuggerDisplayAttribute", symbolicModel.Compilation, generator.LiteralExpression("{" + DebuggerDisplayMethodName + "(),nq}"))
+                .WithAdditionalAnnotations(Simplifier.Annotation);
+
+            var modifiedType = generator.AddAttributes(type, newAttribute);
```

Comment:
you could also do SyntaxEditor.  Then you could both add hte attributes and add the member without needing to keep track of the intermediary edits.  up to you.

Reply:
Never used SyntaxEditor. I'm interested.

Reply:
This turned out nice. Thanks!